src/chapter_3/PDAs.md

```diff
@@ -0,0 +1,315 @@
+When the CPI is invoked, for each account in `cpi_accounts` the Solana runtime will check whether`hash(seeds, current_program_id) == account address` is true. If yes, that account's `is_signer` flag will be turned to true.
+This means a PDA derived from some program X, may only be used to sign CPIs that originate from that program X. This means that on a high level, PDA signatures can be considered program signatures.
+
+This is great news because for many programs it is necessary that the program itself takes the authority over some assets.
+For instance, lending protocol programs need to manage deposited collateral and automated market maker programs need to manage the tokens put into their liquidity pools.
+
+Before we conclude this chapter, let's revisit the puppet workspace and add a PDA signature.
+
+First, adjust the puppet-master code:
+```rust,ignore
+use anchor_lang::prelude::*;
+use puppet::cpi::accounts::SetData;
+use puppet::program::Puppet;
+use puppet::{self, Data};
+
+declare_id!("HmbTLCmaGvZhKnn1Zfa1JVnp7vkMV4DYVxPLWBVoN65L");
+
+#[program]
+mod puppet_master {
+    use super::*;
+    pub fn pull_strings(ctx: Context<PullStrings>, bump: u8, data: u64) -> ProgramResult {
+        let cpi_program = ctx.accounts.puppet_program.to_account_info();
+        let cpi_accounts = SetData {
+            puppet: ctx.accounts.puppet.to_account_info(),
+            authority: ctx.accounts.authority.to_account_info()
+        };
+        let bumps = &[bump][..];
+        let signer = &[&[bumps][..]];
+        let cpi_ctx = CpiContext::new_with_signer(cpi_program, cpi_accounts, signer);
+        puppet::cpi::set_data(cpi_ctx, data)
+    }
+}
+
+#[derive(Accounts)]
+pub struct PullStrings<'info> {
+    #[account(mut)]
+    pub puppet: Account<'info, Data>,
+    pub puppet_program: Program<'info, Puppet>,
+    pub authority: UncheckedAccount<'info>
+}
+```
+
+The `authority` account is now an `UncheckedAccount` instead of a `Signer`. When the puppet-master is invoked, the `authority` pda is not a signer yet so we mustn't add a check for it. We just care about the puppet-master being able to sign so we don't add any additional seeds. Just a bump that is calculated off-chain and then passed to the function. This is the new `puppet.ts`:
+```ts
+import * as anchor from '@project-serum/anchor';
+import { web3 } from '@project-serum/anchor/';
+import { Program } from '@project-serum/anchor';
+import { Puppet } from '../target/types/puppet';
+import { PuppetMaster } from '../target/types/puppet_master';
+import { expect } from 'chai';
+
+describe('puppet', () => {
+  anchor.setProvider(anchor.Provider.env());
+
+  const puppetProgram = anchor.workspace.Puppet as Program<Puppet>;
+  const puppetMasterProgram = anchor.workspace.PuppetMaster as Program<PuppetMaster>;
+
+  const puppetKeypair = web3.Keypair.generate();
```

Comment:
We should import Keypair directly instead of doing web3.Keypair.

Reply:
Similarly, for all the other web3.* types.

Reply:
I tried that but couldnt find a nice way to do it. Is there a better way than:
```ts
import * as anchor from '@project-serum/anchor';
const web3 = anchor.web3;
const Keypair = web3.Keypair;
```

Reply:
`import { Keypair } from '@solana/web3.js'`